pygeos/_geometry.pyx

```diff
@@ -75,3 +79,114 @@ def get_parts(object[:] array):
+    geometries = np.asarray(geometries, dtype=np.object)
+    if geometries.ndim != 1:
+        raise TypeError("geometries is not a one-dimensional array.")
+
+    indices = np.asarray(indices, dtype=np.int32)
+    if indices.ndim != 1:
+        raise TypeError("indices is not a one-dimensional array.")
+
+    if geometries.shape[0] != indices.shape[0]:
+        raise ValueError("geometries and indices do not have equal size.")
+
+    if geometries.shape[0] == 0:
+        # return immediately if there are no geometries to return
+        return np.empty(shape=(0, ), dtype=np.object_)
+
+    if np.any(indices[1:] < indices[:-1]):
+        raise ValueError("The indices should be sorted.")  
+
+    cdef object[:] geometries_view = geometries
+    cdef int[:] indices_view = indices
+
+    # get the geometry count per collection
+    cdef int[:] collection_size = np.bincount(indices).astype(np.int32)
+
+    # A temporary array for the geometries that will be given to CreateCollection.
+    # Its size equals max(collection_size) to accomodate the largest collection.
+    temp_geoms = np.empty(shape=(np.max(collection_size), ), dtype=np.intp)
+    cdef np.intp_t[:] temp_geoms_view = temp_geoms
+
+    # The final target array
+    cdef Py_ssize_t n_colls = collection_size.shape[0]
+    result = np.empty(shape=(n_colls, ), dtype=np.object_)
+    cdef object[:] result_view = result
+
+    with get_geos_handle() as geos_handle:
+        for coll_idx in range(n_colls):
+            coll_size = 0
+
+            # fill the temporary array with geometries belonging to this collection
+            for coll_geom_idx in range(collection_size[coll_idx]):
+                if PyGEOS_GetGEOSGeometry(<PyObject *>geometries_view[geom_idx_1 + coll_geom_idx], &geom) == 0:
+                    # deallocate previous temp geometries (preventing memory leaks)
+                    for coll_geom_idx in range(coll_size):
+                        GEOSGeom_destroy_r(geos_handle, <GEOSGeometry *>temp_geoms_view[coll_geom_idx])
+                    raise TypeError(
+                        "One of the arguments is of incorrect type. Please provide only Geometry objects."
+                    )
+
+                # Check geometry subtype for non-geometrycollections
+                if geometry_type != 7:
+                    curr_type = GEOSGeomTypeId_r(geos_handle, geom)
+                    if curr_type != expected_type and curr_type != expected_type_alt:
+                        # deallocate previous temp geometries (preventing memory leaks)
+                        for coll_geom_idx in range(coll_size):
+                            GEOSGeom_destroy_r(geos_handle, <GEOSGeometry *>temp_geoms_view[coll_geom_idx])
+                        raise TypeError(
+                            f"One of the arguments has unexpected geometry type {curr_type}."
+                        )
+
+                # ignore missing values
+                if geom != NULL:
```

Comment:
The `NULL` check should come above line 163 so that we don't call `GEOSGeomTypeId_r` with a `NULL` geometry.